language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -266,6 +268,65 @@ instantiateEvaluate con ty fun
     | otherwise                      =
         throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
 
+{- Note [Saved mapping example]
+Consider a polymorphic built-in function @id@, whose type signature on the Plutus side is
+
+    id : all a. a -> a
+
+Notation:
+
+- the variable environment is denoted as @{ <var> :-> (<env>, <value>), ... }@
+- the context is denoted as a Haskell list of 'Frame's
+- 'computeCek' is denoted as @(|>)@
+- 'returnCek' is denoted as @(<|)@
+
+When evaluating
+
+    id {integer -> integer} ((\(i : integer) (j : integer) -> i) 1) 0
+
+We encounter the following state:
+
+    { i :-> ({}, 1) }
+    [ FrameApplyFun {} (id {integer -> integer})
+    , FrameApplyArg {} 0
+    ] |> \(j : integer) -> i
+
+and transition it into
+
+    { arg :-> ({ i :-> ({}, 1) }, \(j : integer) -> i) }
+    [ FrameApplyArg {} 0
+    ] <| id {integer -> integer} arg
+
+i.e. if the argument is not a constant, then we create a new variable, save the old environment in
+the closure of that variable and apply the function to the variable. This allows to restore the old
+environment @{ i :-> ({}, 1) }@ latter when we start evaluating @arg 0@, which expands to
+
+    (\(j : integer) -> i)) 0
+
+which evaluates to @1@ in the old environment.
```

Comment:
So if I understand it, the argument is that *variables* are the easiest thing to pass "through" a polymorphic builtin: if it was in scope as an argument, it will be in scope as a result (I guess since we can't introduce any new scopes in between the applied builtin and its arguments).

That gives me an idea: what if we shift the work to the user? i.e. we stipulate that the arguments to builtins must be precisely variables or constants. Then it's up to the compiler to make sure that any non-variable arguments get lifted into variables.

My main concern here is that this restriction only really makes sense when you think of the CEK machine, since the CK machine can do just fine with arbitrary terms as arguments.

Reply:
> So if I understand it, the argument is that variables are the easiest thing to pass "through" a polymorphic builtin: if it was in scope as an argument, it will be in scope as a result (I guess since we can't introduce any new scopes in between the applied builtin and its arguments).

The CEK machine evaluates arguments to a function before trying to reduce the entire application and evaluating an argument variable means instantiating that variable. So even if you eta-expand all built-in functions, that won't suffice. I've dug into this rabbit hole [here](https://github.com/input-output-hk/plutus/issues/1882#issuecomment-598716038).

Reply:
So yes, we could do that provided we

- don't change the AST to express this invariant
- have saturated builtins, so that we can distinguish between an argument to a lambda (which needs to be instantiated) and an argument to a builtin (which doesn't). Not confident this is a must, though

Reply:
Ah right, we do indeed return the value with its environment, which we'd then lose if we dropped the argument environment.